presto-testing-docker/src/main/java/io/prestosql/testing/TemporaryDockerContainer.java

```diff
@@ -0,0 +1,215 @@
+            throws Exception
+    {
+        Map<String, List<PortBinding>> portBindings = ports.stream()
+                .collect(toImmutableMap(Object::toString, port -> ImmutableList.of(PortBinding.create(hostIp, "0"))));
+        String[] exposedPorts = ports.stream()
+                .map(Object::toString)
+                .collect(toImmutableList())
+                .toArray(new String[ports.size()]);
+
+        containerId = dockerClient.createContainer(ContainerConfig.builder()
+                .hostConfig(HostConfig.builder()
+                        .portBindings(portBindings)
+                        .build())
+                .exposedPorts(exposedPorts)
+                .env(environment.entrySet().stream()
+                        .map(entry -> format("%s=%s", entry.getKey(), entry.getValue()))
+                        .collect(toImmutableList()))
+                .hostname(hostName)
+                .image(image)
+                .build())
+                .id();
+
+        LOG.info("Started docker container with id: %s", containerId);
+
+        dockerClient.startContainer(containerId);
+
+        calculateHostPorts(ports);
+
+        ports.forEach(port -> {
+            RetryPolicy<Object> retryPolicy = new RetryPolicy<>()
+                    .withMaxDuration(Duration.of(5, MINUTES))
+                    .withMaxAttempts(Integer.MAX_VALUE) // retrying for 5 minutes as many times as possible, once per 5 second
+                    .withDelay(Duration.of(5, SECONDS))
+                    .onRetry(event -> LOG.info("Waiting for port %s that is exposed on host as %s:%s ...", port, hostIp, getHostPort(port)));
+            Failsafe.with(retryPolicy).run(() -> {
+                try (Socket socket = new Socket(hostIp, getHostPort(port))) {
+                    checkState(socket.isConnected());
+                }
+            });
+        });
+    }
+
+    private void calculateHostPorts(List<Integer> ports)
+            throws Exception
+    {
+        hostPorts = dockerClient.inspectContainer(containerId).networkSettings()
+                .ports()
+                .entrySet()
+                .stream()
+                .filter(entry -> ports.contains(extractPort(entry)))
+                .collect(toImmutableMap(
+                        entry -> extractPort(entry),
+                        entry -> entry.getValue().stream()
+                                .filter(portBinding -> portBinding.hostIp().equals(hostIp))
+                                .map(PortBinding::hostPort)
+                                .collect(toOptional())
+                                .map(Integer::parseInt)
+                                .orElseThrow(() -> new IllegalStateException("Could not find port mapping for: " + entry.getKey()))));
+    }
+
+    private int extractPort(Map.Entry<String, List<PortBinding>> entry)
+    {
+        return Integer.parseInt(entry.getKey().replace("/tcp", "").replace("/udp", ""));
```

Comment:
AFAIK, Docker for Mac does not handle UDP correctly, so let's not pretend this is supported.

Reply:
What happens when eg 88/tcp and 88/udp are bound but mapped to a different host port?